File: backend/routes/security.js
const express = require('express');
const router = express.Router();

// In-memory event store — TODO: Replace with PostgreSQL
const securityEvents = [];
const MAX_EVENTS = 10000;

/**
 * POST /api/security/events
 * Logs a security event from the Android app.
 *
 * Body: {
 *   type: string,       // e.g., "THREAT_DETECTED", "VPN_CONNECTED", "SCAN_COMPLETED", "SOS_TRIGGERED"
 *   severity: string,   // "LOW", "MEDIUM", "HIGH", "CRITICAL"
 *   source: string,     // e.g., "url_scanner", "vpn", "sound_detector", "camera"
 *   details: object,    // Event-specific data
 *   deviceId?: string,
 *   userId?: string
 * }
 */
router.post('/events', (req, res) => {
    const { type, severity = 'LOW', source, details = {}, deviceId, userId } = req.body;

    if (!type) return res.status(400).json({ error: 'type is required' });
    if (!source) return res.status(400).json({ error: 'source is required' });

    const validSeverities = ['LOW', 'MEDIUM', 'HIGH', 'CRITICAL'];
    if (!validSeverities.includes(severity.toUpperCase())) {
        return res.status(400).json({ error: `severity must be one of: ${validSeverities.join(', ')}` });
    }

    const event = {
        id: `EVT-${Date.now()}-${Math.random().toString(36).substring(2, 8)}`,
        type: type.toUpperCase(),
        severity: severity.toUpperCase(),
        source,
        details,
        deviceId: deviceId || 'unknown',
        userId: userId || 'anonymous',
        timestamp: new Date().toISOString(),
        ip: req.ip
    };

    // Keep events within limit
    if (securityEvents.length >= MAX_EVENTS) {
        securityEvents.shift(); // Remove oldest
    }
    securityEvents.push(event);

    console.log(`[Security] ${event.severity} | ${event.type} from ${event.source} | device=${event.deviceId}`);

    res.status(201).json({
        success: true,
        eventId: event.id,
        message: 'Security event logged'
    });
});

/**
 * GET /api/security/events
 * Returns recent security events.
 * Query params: ?limit=50&severity=HIGH&type=THREAT_DETECTED&source=url_scanner
 */
router.get('/events', (req, res) => {
    const internalSecret = process.env.API_INTERNAL_SECRET;
    const providedSecret = req.headers['x-internal-secret'];

    // Protect read access with internal secret
    if (internalSecret && providedSecret !== internalSecret) {
        return res.status(403).json({ error: 'Forbidden — internal secret required' });
    }

    let filtered = [...securityEvents];

    // Filter by severity
    if (req.query.severity) {
        filtered = filtered.filter(e => e.severity === req.query.severity.toUpperCase());
    }

    // Filter by type
    if (req.query.type) {
        filtered = filtered.filter(e => e.type === req.query.type.toUpperCase());
    }

    // Filter by source
    if (req.query.source) {
        filtered = filtered.filter(e => e.source === req.query.source);
    }

    // Filter by deviceId
    if (req.query.deviceId) {
        filtered = filtered.filter(e => e.deviceId === req.query.deviceId);
    }

    // Limit and sort (newest first)
    const limit = Math.min(parseInt(req.query.limit) || 50, 500);
    filtered = filtered.reverse().slice(0, limit);

    res.json({
        success: true,
        total: securityEvents.length,
        returned: filtered.length,
        events: filtered
    });
});

/**
 * GET /api/security/summary
 * Returns a summary of security events
 */
router.get('/summary', (req, res) => {
    const last24h = new Date(Date.now() - 24 * 60 * 60 * 1000);
    const recent = securityEvents.filter(e => new Date(e.timestamp) > last24h);

    const bySeverity = { LOW: 0, MEDIUM: 0, HIGH: 0, CRITICAL: 0 };
    const byType = {};
    const bySource = {};

    recent.forEach(e => {
        bySeverity[e.severity] = (bySeverity[e.severity] || 0) + 1;
        byType[e.type] = (byType[e.type] || 0) + 1;
        bySource[e.source] = (bySource[e.source] || 0) + 1;
    });

    res.json({
        success: true,
        period: '24h',
        totalEvents: recent.length,
        bySeverity,
        byType,
        bySource
    });
});

module.exports = router;

File: royal_shield_backend_unified/api/routes/security.py
from fastapi import APIRouter, HTTPException, status, Header, Request
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
import logging
import time
import uuid
from datetime import datetime, timedelta
from config.settings import settings

router = APIRouter(prefix="/api/security", tags=["Security"])
logger = logging.getLogger(__name__)

# In-memory event store
SECURITY_EVENTS = []
MAX_EVENTS = 10000

class EventPayload(BaseModel):
    type: str
    severity: str = "LOW"
    source: str
    details: Optional[Dict[str, Any]] = Field(default_factory=dict)
    deviceId: Optional[str] = None
    userId: Optional[str] = None

@router.post("/events", status_code=status.HTTP_201_CREATED)
def log_security_event(payload: EventPayload, request: Request):
    if not payload.type:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="type is required")
    if not payload.source:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="source is required")

    valid_severities = ['LOW', 'MEDIUM', 'HIGH', 'CRITICAL']
    severity_upper = payload.severity.upper()
    if severity_upper not in valid_severities:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"severity must be one of: {', '.join(valid_severities)}"
        )

    event_id = f"EVT-{int(time.time() * 1000)}-{str(uuid.uuid4())[:6]}"
    ip_address = request.client.host if request.client else "unknown"

    event = {
        "id": event_id,
        "type": payload.type.upper(),
        "severity": severity_upper,
        "source": payload.source,
        "details": payload.details,
        "deviceId": payload.deviceId or "unknown",
        "userId": payload.userId or "anonymous",
        "timestamp": datetime.utcnow().isoformat() + "Z",
        "ip": ip_address
    }

    # Lock & Append
    if len(SECURITY_EVENTS) >= MAX_EVENTS:
        SECURITY_EVENTS.pop(0)  # Remove oldest

    SECURITY_EVENTS.append(event)
    logger.info(f"[Security] {event['severity']} | {event['type']} from {event['source']} | device={event['deviceId']}")

    return {
        "success": True,
        "eventId": event_id,
        "message": "Security event logged"
    }

@router.get("/events")
def get_security_events(
    x_internal_secret: Optional[str] = Header(None),
    limit: int = 50,
    severity: Optional[str] = None,
    type: Optional[str] = None,
    source: Optional[str] = None,
    deviceId: Optional[str] = None
):
    internal_secret = settings.api_internal_secret

    if internal_secret and x_internal_secret != internal_secret:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Forbidden — internal secret required"
        )

    filtered = list(SECURITY_EVENTS)

    if severity:
        filtered = [e for e in filtered if e["severity"] == severity.upper()]
    if type:
        filtered = [e for e in filtered if e["type"] == type.upper()]
    if source:
        filtered = [e for e in filtered if e["source"] == source]
    if deviceId:
        filtered = [e for e in filtered if e["deviceId"] == deviceId]

    # Limit and sort (newest first)
    limit = min(max(limit, 1), 500)
    filtered.reverse()
    result_slice = filtered[:limit]

    return {
        "success": True,
        "total": len(SECURITY_EVENTS),
        "returned": len(result_slice),
        "events": result_slice
    }

@router.get("/summary")
def get_security_summary():
    last_24h = datetime.utcnow() - timedelta(hours=24)
    recent = []

    for e in SECURITY_EVENTS:
        try:
            # Parse timestamp (stripping trailing Z for datetime.fromisoformat)
            ts_str = e["timestamp"]
            if ts_str.endswith("Z"):
                ts_str = ts_str[:-1]
            ts = datetime.fromisoformat(ts_str)
            if ts > last_24h:
                recent.append(e)
        except Exception:
            pass

    by_severity = {"LOW": 0, "MEDIUM": 0, "HIGH": 0, "CRITICAL": 0}
    by_type = {}
    by_source = {}

    for e in recent:
        sev = e["severity"]
        by_severity[sev] = by_severity.get(sev, 0) + 1

        t = e["type"]
        by_type[t] = by_type.get(t, 0) + 1

        s = e["source"]
        by_source[s] = by_source.get(s, 0) + 1

    return {
        "success": True,
        "period": "24h",
        "totalEvents": len(recent),
        "bySeverity": by_severity,
        "byType": by_type,
        "bySource": by_source
    }
